File: gqlkit-ts/src/builder.ts
/**
 * @module builder
 *
 * Provides the query-building primitives that generated SDK code extends.
 * - {@link FieldSelection} models the recursive field tree in a GraphQL selection set.
 * - {@link BaseBuilder} assembles a complete operation string (with variables) and executes it.
 */

import { GraphQLClient } from "./graphqlclient";

/**
 * Tracks which fields are selected in a GraphQL query.
 *
 * Maintains two lists:
 * - `fields`   : Scalar (leaf) field names like `id`, `name`.
 * - `children` : Nested object selections, each with its own {@link FieldSelection}.
 *
 * Used by generated builder classes to accumulate the selection set before
 * serializing it into a GraphQL query string via {@link build}.
 *
 * @example
 * ```ts
 * const sel = new FieldSelection();
 * sel.addField("id");
 * sel.addField("name");
 * const addressSel = new FieldSelection();
 * addressSel.addField("city");
 * sel.addChild("address", addressSel);
 * console.log(sel.build());
 * // =>
 * //   id
 * //   name
 * //   address {
 * //     city
 * //   }
 * ```
 */
export class FieldSelection {
  /** Scalar field names at this level of the selection. */
  private fields: string[] = [];

  /** Nested object fields, each mapping a field name to its sub-selection. */
  private children: Map<string, FieldSelection> = new Map();

  /**
   * Add a scalar (leaf) field to this selection level.
   * @param name - The field name to include (e.g., "id", "email").
   */
  addField(name: string): void {
    this.fields.push(name);
  }

  /**
   * Add a nested object field with its own sub-selection.
   * @param name  - The field name that returns an object type.
   * @param child - The {@link FieldSelection} describing which sub-fields to request.
   */
  addChild(name: string, child: FieldSelection): void {
    this.children.set(name, child);
  }

  /**
   * Serialize this selection into a GraphQL selection-set string.
   * Recursively builds nested selections with increasing indentation.
   *
   * @param indent - Number of leading spaces for the current depth (default: 2).
   * @returns A multi-line string representing the GraphQL field selection.
   */
  build(indent: number = 2): string {
    const pad = " ".repeat(indent);
    const parts: string[] = [];

    // Emit each scalar field on its own indented line
    for (const field of this.fields) {
      parts.push(`${pad}${field}`);
    }

    // Emit each nested field with its sub-selection wrapped in braces
    for (const [name, child] of this.children) {
      const nested = child.build(indent + 2);
      parts.push(`${pad}${name} {\n${nested}\n${pad}}`);
    }

    return parts.join("\n");
  }

  /**
   * Check whether this selection contains any fields at all.
   * An empty selection means no fields were requested at this level.
   * @returns `true` if no scalar fields and no nested children exist.
   */
  isEmpty(): boolean {
    return this.fields.length === 0 && this.children.size === 0;
  }
}

/**
 * Base class for generated GraphQL operation builders (queries and mutations).
 *
 * Generated SDK classes extend BaseBuilder to provide typed setter methods for
 * each argument and field-selection helpers. BaseBuilder handles:
 * - Storing operation arguments with their GraphQL type annotations.
 * - Accumulating the field selection via {@link FieldSelection}.
 * - Serializing everything into a valid GraphQL operation string.
 * - Executing the operation through the provided {@link GraphQLClient}.
 *
 * A typical generated builder calls `setArg()` for user-supplied arguments,
 * populates the selection via `getSelection()`, and finally calls `executeRaw()`
 * to send the request.
 *
 * @example
 * ```
 * // Generated code (simplified):
 * class GetUserBuilder extends BaseBuilder {
 *   constructor(client: GraphQLClient) {
 *     super(client, "query", "GetUser", "user");
 *   }
 *   id(value: string) { this.setArg("id", value, "ID!"); return this; }
 *   async exec() { return (await this.executeRaw()).user as User; }
 * }
 * ```
 */
export class BaseBuilder {
  /** The GraphQL client used to execute the built operation. */
  private client: GraphQLClient;

  /** The operation type keyword: "query" or "mutation". */
  private opType: string;

  /** The operation name used in the GraphQL document (e.g., "GetUser"). */
  private opName: string;

  /** The root field name inside the operation (e.g., "user", "createUser"). */
  private fieldName: string;

  /**
   * Map of argument name to its runtime value and GraphQL type string.
   * For example: "id" -> { value: "123", graphqlType: "ID!" }
   */
  private args: Map<string, { value: unknown; graphqlType: string }> =
    new Map();

  /** The field selection tree for the operation's return type. */
  private selection: FieldSelection = new FieldSelection();

  /**
   * @param client    - The GraphQL client to execute operations with.
   * @param opType    - Operation kind: "query" or "mutation".
   * @param opName    - The named operation identifier (e.g., "GetUser").
   * @param fieldName - The root field to query/mutate (e.g., "user").
   */
  constructor(
    client: GraphQLClient,
    opType: string,
    opName: string,
    fieldName: string
  ) {
    this.client = client;
    this.opType = opType;
    this.opName = opName;
    this.fieldName = fieldName;
  }

  /**
   * Register an argument for the operation.
   * Called by generated setter methods to record each argument's value and type.
   *
   * @param name        - The argument name (doubles as the variable name).
   * @param value       - The runtime value to send as a variable.
   * @param graphqlType - The GraphQL type annotation (e.g., "String!", "Int", "[ID!]!").
   */
  setArg(name: string, value: unknown, graphqlType: string): void {
    this.args.set(name, { value, graphqlType });
  }

  /**
   * Access the field selection tree so callers can add fields and nested selections.
   * @returns The root {@link FieldSelection} for this operation.
   */
  getSelection(): FieldSelection {
    return this.selection;
  }

  /**
   * Access the underlying GraphQL client (used by generated code if needed).
   * @returns The {@link GraphQLClient} instance.
   */
  getClient(): GraphQLClient {
    return this.client;
  }

  /**
   * Extract the runtime variable values as a plain object for the request payload.
   * Maps each registered argument name to its value, discarding the type info.
   *
   * @returns A `{ [argName]: value }` record suitable for the `variables` field.
   */
  getVariables(): Record<string, unknown> {
    const vars: Record<string, unknown> = {};
    for (const [name, { value }] of this.args) {
      vars[name] = value;
    }
    return vars;
  }

  /**
   * Build the complete GraphQL operation string.
   *
   * Produces output like:
   * ```graphql
   * query GetUser($id: ID!) {
   *   user(id: $id) {
   *     id
   *     name
   *   }
   * }
   * ```
   *
   * Steps:
   * 1. Build variable declarations from registered args (e.g., `$id: ID!`).
   * 2. Build argument pass-throughs for the root field (e.g., `id: $id`).
   * 3. Serialize the field selection set (if any fields were selected).
   * 4. Combine into the final operation string.
   *
   * @returns The full GraphQL operation document as a string.
   */
  buildQuery(): string {
    // Collect variable declarations (e.g., "$id: ID!") and argument references (e.g., "id: $id")
    const varDecls: string[] = [];
    const argPasses: string[] = [];

    for (const [name, { graphqlType }] of this.args) {
      varDecls.push(`$${name}: ${graphqlType}`);
      argPasses.push(`${name}: $${name}`);
    }

    // Format the variable declaration and argument lists (empty string if no args)
    const varStr = varDecls.length > 0 ? `(${varDecls.join(", ")})` : "";
    const argStr = argPasses.length > 0 ? `(${argPasses.join(", ")})` : "";

    // Build the selection set body; omit braces entirely if no fields selected
    const selStr = this.selection.isEmpty()
      ? ""
      : ` {\n${this.selection.build(4)}\n  }`;

    // Assemble: "query GetUser($id: ID!) {\n  user(id: $id) {\n    ...\n  }\n}"
    return `${this.opType} ${this.opName}${varStr} {\n  ${this.fieldName}${argStr}${selStr}\n}`;
  }

  /**
   * Build and execute the operation, returning the raw `data` object.
   *
   * Generated subclasses typically wrap this method to extract and cast
   * the specific root field from the response (e.g., `(await executeRaw()).user`).
   *
   * @returns The full `data` object from the GraphQL response.
   * @throws {GraphQLErrors} If the server returns errors.
   */
  async executeRaw(): Promise<Record<string, unknown>> {
    const query = this.buildQuery();
    const variables = this.getVariables();
    return await this.client.execute<Record<string, unknown>>(
      query,
      variables
    );
  }

  /**
   * Produce the pieces needed to merge this operation into a batched, multi-root
   * operation under a caller-chosen alias. Argument names are namespaced with
   * the alias so multiple builders sharing argument names (e.g., two `tasks`
   * calls each taking `$status`) do not collide in the merged document.
   *
   * @param alias - GraphQL response key for this operation's root field.
   * @returns Operation kind, prefixed variable declarations + values, and the
   *          aliased root-field string ready to be concatenated into a batch.
   */
  getOpFragment(alias: string): {
    opType: string;
    varDecls: string[];
    varValues: Record<string, unknown>;
    aliasedField: string;
  } {
    const varDecls: string[] = [];
    const argPasses: string[] = [];
    const varValues: Record<string, unknown> = {};

    for (const [name, { value, graphqlType }] of this.args) {
      const prefixed = `${alias}_${name}`;
      varDecls.push(`$${prefixed}: ${graphqlType}`);
      argPasses.push(`${name}: $${prefixed}`);
      varValues[prefixed] = value;
    }

    const argStr = argPasses.length > 0 ? `(${argPasses.join(", ")})` : "";
    const selStr = this.selection.isEmpty()
      ? ""
      : ` {\n${this.selection.build(4)}\n  }`;

    const aliasedField = `${alias}: ${this.fieldName}${argStr}${selStr}`;

    return { opType: this.opType, varDecls, varValues, aliasedField };
  }
}

File: gqlkit/pkg/builder/builder.go
// Package builder provides the runtime foundation for generated GraphQL operation
// builders. It contains FieldSelection (a recursive tree that assembles the
// requested fields of a GraphQL selection set) and BaseBuilder (the shared logic
// every generated query/mutation builder delegates to for argument tracking,
// query string construction, and execution).
package builder

import (
	"context"
	"sort"
	"strings"
)

// GraphQLClient is the interface that the generated SDK expects consumers to
// implement (or satisfy via graphqlclient.Client). It sends a single GraphQL
// request and unmarshals the JSON response into the supplied target.
type GraphQLClient interface {
	Execute(ctx context.Context, query string, variables map[string]interface{}, response any) error
}

// FieldSelection tracks selected fields for a GraphQL query.
// Scalar fields are stored in `fields`; nested object selections are stored in
// `children`, forming a recursive tree that mirrors the GraphQL selection set.
type FieldSelection struct {
	fields   []string                    // leaf/scalar field names
	children map[string]*FieldSelection  // nested object selections keyed by field name
}

// NewFieldSelection creates a new FieldSelection
func NewFieldSelection() *FieldSelection {
	return &FieldSelection{
		fields:   make([]string, 0),
		children: make(map[string]*FieldSelection),
	}
}

// AddField adds a scalar field to the selection
func (fs *FieldSelection) AddField(name string) {
	fs.fields = append(fs.fields, name)
}

// AddChild adds a nested field selection
func (fs *FieldSelection) AddChild(name string, child *FieldSelection) {
	fs.children[name] = child
}

// Build recursively renders the field selection into a GraphQL selection set
// string. Each level is indented by `indent` levels of two spaces. Child keys
// are sorted alphabetically to ensure deterministic output across runs.
func (fs *FieldSelection) Build(indent int) string {
	if len(fs.fields) == 0 && len(fs.children) == 0 {
		return ""
	}

	var sb strings.Builder
	prefix := strings.Repeat("  ", indent)

	for _, field := range fs.fields {
		sb.WriteString(prefix + field + "\n")
	}

	// Sort children keys for deterministic output
	childKeys := make([]string, 0, len(fs.children))
	for k := range fs.children {
		childKeys = append(childKeys, k)
	}
	sort.Strings(childKeys)

	for _, name := range childKeys {
		child := fs.children[name]
		childStr := child.Build(indent + 1)
		if childStr != "" {
			sb.WriteString(prefix + name + " {\n")
			sb.WriteString(childStr)
			sb.WriteString(prefix + "}\n")
		}
	}

	return sb.String()
}

// BaseBuilder provides the shared state and logic for every generated operation
// builder. It tracks the GraphQL client, the operation type ("query" or
// "mutation"), the operation and field names, arguments with their GraphQL type
// annotations, and the field selection tree. Generated per-operation builders
// embed *BaseBuilder and add typed setter methods.
type BaseBuilder struct {
	client    GraphQLClient              // underlying HTTP transport
	opType    string                     // "query" or "mutation"
	opName    string                     // PascalCase operation name used in the query string
	fieldName string                     // the root field name inside the operation
	args      map[string]interface{}     // argument values keyed by GraphQL arg name
	argTypes  map[string]string          // GraphQL type strings keyed by arg name (e.g. "Int!", "[ID!]")
	selection *FieldSelection            // selected fields for the response
}

// NewBaseBuilder creates a new BaseBuilder
func NewBaseBuilder(client GraphQLClient, opType, opName, fieldName string) *BaseBuilder {
	return &BaseBuilder{
		client:    client,
		opType:    opType,
		opName:    opName,
		fieldName: fieldName,
		args:      make(map[string]interface{}),
		argTypes:  make(map[string]string),
		selection: NewFieldSelection(),
	}
}

// SetArg records an argument value and its corresponding GraphQL type string.
// Generated builders call this from their typed setter methods (e.g. SetLimit).
func (b *BaseBuilder) SetArg(name string, value interface{}, graphqlType string) {
	b.args[name] = value
	b.argTypes[name] = graphqlType
}

// GetSelection returns the field selection
func (b *BaseBuilder) GetSelection() *FieldSelection {
	return b.selection
}

// GetClient returns the client
func (b *BaseBuilder) GetClient() GraphQLClient {
	return b.client
}

// BuildQuery assembles the full GraphQL query/mutation string from the
// operation metadata, registered arguments, and field selection tree.
// Variable declarations are sorted alphabetically for deterministic output.
func (b *BaseBuilder) BuildQuery() string {
	var sb strings.Builder

	sb.WriteString(b.opType + " " + b.opName)

	if len(b.args) > 0 {
		sb.WriteString("(")
		vars := make([]string, 0, len(b.args))
		for name, gqlType := range b.argTypes {
			vars = append(vars, "$"+name+": "+gqlType)
		}
		sort.Strings(vars)
		sb.WriteString(strings.Join(vars, ", "))
		sb.WriteString(")")
	}

	sb.WriteString(" {\n")
	sb.WriteString("  " + b.fieldName)

	if len(b.args) > 0 {
		sb.WriteString("(")
		args := make([]string, 0, len(b.args))
		for name := range b.args {
			args = append(args, name+": $"+name)
		}
		sort.Strings(args)
		sb.WriteString(strings.Join(args, ", "))
		sb.WriteString(")")
	}

	selectionStr := b.selection.Build(2)
	if selectionStr != "" {
		sb.WriteString(" {\n")
		sb.WriteString(selectionStr)
		sb.WriteString("  }")
	}

	sb.WriteString("\n}")
	return sb.String()
}

// GetVariables returns the variables map
func (b *BaseBuilder) GetVariables() map[string]interface{} {
	return b.args
}

// ExecuteRaw builds the query, sends it via the GraphQL client, and returns the
// top-level response data as an untyped map. This is used internally by generated
// builders before unmarshalling into the concrete response struct.
func (b *BaseBuilder) ExecuteRaw(ctx context.Context) (map[string]interface{}, error) {
	query := b.BuildQuery()
	variables := b.GetVariables()

	var response map[string]interface{}
	if err := b.client.Execute(ctx, query, variables, &response); err != nil {
		return nil, err
	}

	return response, nil
}

// OpFragment is the slice of an operation produced by a single builder when it
// is being merged into a batched, multi-root document. The TypeScript runtime
// has the same shape — see gqlkit-ts/src/batch.ts.
//
// VarDecls and VarValues use the alias-prefixed variable names (e.g.
// "$open_filter" instead of "$filter") so two builders sharing an argument
// name can coexist in the merged operation.
type OpFragment struct {
	OpType       string
	VarDecls     []string
	VarValues    map[string]interface{}
	AliasedField string
}

// GetOpFragment renders this builder as a fragment that batch.RunQueries /
// batch.RunMutations can splice into a single GraphQL document. The returned
// AliasedField looks like:
//
//	open: todos(filter: $open_filter) {
//	    id
//	    text
//	  }
//
// Argument names are prefixed with the supplied alias to prevent variable
// collisions when merging multiple builders.
func (b *BaseBuilder) GetOpFragment(alias string) OpFragment {
	frag := OpFragment{
		OpType:    b.opType,
		VarValues: make(map[string]interface{}, len(b.args)),
	}

	// Sort argument names for deterministic output across builds
	argNames := make([]string, 0, len(b.args))
	for name := range b.args {
		argNames = append(argNames, name)
	}
	sort.Strings(argNames)

	argPasses := make([]string, 0, len(b.args))
	for _, name := range argNames {
		prefixed := alias + "_" + name
		gqlType := b.argTypes[name]
		frag.VarDecls = append(frag.VarDecls, "$"+prefixed+": "+gqlType)
		argPasses = append(argPasses, name+": $"+prefixed)
		frag.VarValues[prefixed] = b.args[name]
	}

	var sb strings.Builder
	sb.WriteString(alias + ": " + b.fieldName)
	if len(argPasses) > 0 {
		sb.WriteString("(")
		sb.WriteString(strings.Join(argPasses, ", "))
		sb.WriteString(")")
	}

	selectionStr := b.selection.Build(2)
	if selectionStr != "" {
		sb.WriteString(" {\n")
		sb.WriteString(selectionStr)
		sb.WriteString("  }")
	}

	frag.AliasedField = sb.String()
	return frag
}

// QueryMarker is a zero-size embed that tags a generated builder as a
// "query" operation. Embedded into every generated query builder via the
// operation_builder.tmpl template; mutation builders embed MutationMarker.
//
// The marker exists to give the type system something to discriminate on:
// pkg/batch.QueryBatchable requires both GetOpFragment and IsQueryOp, and
// MutationBatchable requires GetOpFragment and IsMutationOp. A query
// builder satisfies the first interface but not the second, so passing a
// mutation builder to batch.RunQueries fails at compile time rather than
// being caught by a runtime op-type check.
//
// We can't gate the interface on op-type alone (every builder embeds the
// same *BaseBuilder), and we can't use generics for this because the input
// is a heterogeneous map. The marker pattern is the cleanest way to lift
// "this is a query" into Go's type system at zero runtime cost.
type QueryMarker struct{}

// IsQueryOp is the marker method that identifies a builder as a query
// operation. It exists solely to satisfy the QueryBatchable interface in
// pkg/batch and has no runtime behaviour.
func (QueryMarker) IsQueryOp() {}

// MutationMarker is the dual of QueryMarker for mutation operations. See
// QueryMarker for the full rationale of the marker pattern.
type MutationMarker struct{}

// IsMutationOp is the marker method that identifies a builder as a mutation
// operation, dual of QueryMarker.IsQueryOp.
func (MutationMarker) IsMutationOp() {}
